client/client.go

```diff
@@ -0,0 +1,227 @@
+	prometheus.MustRegister(pushErrorCounter, pollErrorCounter, scrapeErrorCounter)
+}
+
+func DefaultBackoff() backoff.BackOff {
+	b := backoff.NewExponentialBackOff()
+	b.InitialInterval = 1 * time.Second
+	b.Multiplier = 1.5
+	b.MaxInterval = 5 * time.Second
+	b.MaxElapsedTime = time.Duration(0)
+	return b
+}
+
+// Coordinator for scrape requests and responses
+type Coordinator struct {
+	logger   *slog.Logger
+	client   *http.Client
+	bo       backoff.BackOff
+	fqdn     string
+	proxyUrl string
+}
+
+func NewCoordinator(logger *slog.Logger, bo backoff.BackOff, client *http.Client, fqdn, proxyURL string) (*Coordinator, error) {
+	if fqdn == "" {
+		return nil, errors.New("fqdn must be specified")
+	}
+	if proxyURL == "" {
+		return nil, errors.New("proxyURL must be specified")
+	}
+	if bo == nil {
+		logger.Warn("No backoff provided, using default")
+		bo = DefaultBackoff()
+	}
+	c := &Coordinator{
+		logger:   logger,
+		client:   client,
+		bo:       bo,
+		fqdn:     fqdn,
+		proxyUrl: proxyURL,
+	}
+	return c, nil
+}
+
+func (c *Coordinator) Start(ctx context.Context) {
+	c.loop(ctx)
+}
+
+func (c *Coordinator) handleErr(request *http.Request, err error) {
+	c.logger.Error("Coordinator error", "error", err)
+	scrapeErrorCounter.Inc()
+	resp := &http.Response{
+		StatusCode: http.StatusInternalServerError,
+		Body:       io.NopCloser(strings.NewReader(err.Error())),
+		Header:     http.Header{},
+	}
+	if err = c.doPush(resp, request); err != nil {
+		pushErrorCounter.Inc()
+		c.logger.Warn("Failed to push failed scrape response:", "err", err)
+		return
+	}
+	c.logger.Info("Pushed failed scrape response")
+}
+
+func (c *Coordinator) doScrape(request *http.Request) {
+	logger := c.logger.With("scrape_id", request.Header.Get("id"))
+	timeout, err := util.GetHeaderTimeout(request.Header)
+	if err != nil {
+		c.handleErr(request, err)
+		return
+	}
+	ctx, cancel := context.WithTimeout(request.Context(), timeout)
+	defer cancel()
+	request = request.WithContext(ctx)
+	// We cannot handle https requests at the proxy, as we would only
+	// see a CONNECT, so use a URL parameter to trigger it.
+	params := request.URL.Query()
+	if params.Get("_scheme") == "https" {
+		request.URL.Scheme = "https"
+		params.Del("_scheme")
+		request.URL.RawQuery = params.Encode()
+	}
+
+	if request.URL.Hostname() != c.fqdn {
+		c.handleErr(request, errors.New("scrape target doesn't match client fqdn"))
+		return
+	}
+
+	scrapeResp, err := c.client.Do(request)
+	if err != nil {
+		c.handleErr(request, fmt.Errorf("failed to scrape %s: %w", request.URL.String(), err))
+		return
+	}
+	logger.Info("Retrieved scrape response")
+	if err = c.doPush(scrapeResp, request); err != nil {
+		pushErrorCounter.Inc()
+		logger.Warn("Failed to push scrape response:", "err", err)
+		return
+	}
+	logger.Info("Pushed scrape result")
+}
+
+// Report the result of the scrape back up to the proxy.
+func (c *Coordinator) doPush(resp *http.Response, origRequest *http.Request) error {
```

Comment:
if the comment is correct, why not name the function, reportScrape? or reportProxyScrape?